### DISCUSSION TITLE
Question about sources of WithPerceptionTracking

### DISCUSSION THREAD

**Discussion by Nikolaiko** [pointfreeco/swift-composable-architecture#2772] [Q&A] [answered]
Hi, it's my first post here I don't know is it okay to ask about sources and under tht hood mechanics. Sorry if I doing something wrong.

I just trying to understand how things work and looked into WithPerceptionTracking. And found a body property defined like this:

<img width="776" alt="Снимок экрана 2024-02-05 в 12 18 47" src="https://github.com/pointfreeco/swift-composable-architecture/assets/3236822/0abd09a5-4bb6-4186-96ac-3994444cbc5a">

I think I understand 'else' part. If not iOS 17 we use **withPerceptionTracking** and inside depending on iOS version using **withObservationTracking** or **Perception** backport.

But why if iOS 17 we just return instrumentalBody? Because we assume that in that case macros **@ObservableState** was used on state and it will be work well togther, we don't need to add anyting here?

**Comment by mbrandonw** [comment, marked as answer]
Hi @Nikolaiko, in the `if` branch we see that the user is running iOS 17+, and so SwiftUI does all the heavy lifting for observation. We don't have to do any additional work. Does that help?

**Reply by Nikolaiko** [reply]
@mbrandonw thank you for response!
Yes, that helps.
